loadbalancer/consul/publisher.go

```diff
@@ -0,0 +1,175 @@
+package consul
+
+import (
+	"fmt"
+
+	"github.com/go-kit/kit/endpoint"
+	"github.com/go-kit/kit/loadbalancer"
+	"github.com/go-kit/kit/log"
+	consul "github.com/hashicorp/consul/api"
+)
+
+const defaultIndex = 0
+
+// Publisher yields endpoints for the given service filtered by tags.
+type Publisher struct {
+	cache      *loadbalancer.EndpointCache
+	client     Client
+	logger     log.Logger
+	service    string
+	tags       []string
+	endpointsc chan []endpoint.Endpoint
+	quitc      chan struct{}
+}
+
+// NewPublisher returns a Consul publisher which returns Endpoints for the
+// requested service. It only returns instances for which all of the passed
+// tags are present.
+func NewPublisher(
+	client Client,
+	factory loadbalancer.Factory,
+	logger log.Logger,
+	service string,
+	tags ...string,
+) (*Publisher, error) {
+	logger = log.NewContext(logger).With("component", "Consul Publisher")
+
+	p := &Publisher{
+		cache:   loadbalancer.NewEndpointCache(factory, logger),
+		client:  client,
+		logger:  logger,
+		service: service,
+		tags:    tags,
+		quitc:   make(chan struct{}),
+	}
+
+	is, index, err := p.getInstances(defaultIndex)
+	if err != nil {
+		return nil, err
+	}
+
+	p.cache.Replace(is)
+
+	go p.loop(index)
+
+	return p, nil
+}
+
+// Endpoints implements the Publisher interface.
+func (p *Publisher) Endpoints() ([]endpoint.Endpoint, error) {
+	return p.cache.Endpoints(), nil
+}
+
+// Stop terminates the publisher.
+func (p *Publisher) Stop() {
+	close(p.quitc)
+}
+
+func (p *Publisher) loop(
+	lastIndex uint64,
+) {
+	var (
+		errc = make(chan error)
+		resc = make(chan response)
+	)
+
+	for {
+		go func() {
```

Comment:
I always assume it's safe to push that responsibility to the client [configuration](https://godoc.org/github.com/hashicorp/consul/api#Config) and have the control left with the caller.